src/Libraries/RevitNodes/Application/Document.cs

```diff
@@ -93,6 +108,173 @@ internal ModelPath WorksharingModelPath
             get { return this.InternalDocument.GetWorksharingCentralModelPath(); }
         }
 
+        /// <summary>
+        /// Purge unused Elements from the model.
+        /// </summary>
+        /// <param name="fullPurge">Similar to clicking the purge button multiple times in Revit, the node will repeatedly purge until there is nothing left to remove.</param>
+        /// <returns>Purged element ids</returns>
+        public List<int> PurgeUnused(bool fullPurge = false)
+        {
+            List<ElementId> purgedElementIds = new List<ElementId>();
+            TransactionManager.Instance.EnsureInTransaction(this.InternalDocument);
+
+            purgedElementIds.AddRange(PurgeElements(fullPurge));
+            purgedElementIds.AddRange(PurgeMaterials());
+
+            TransactionManager.Instance.TransactionTaskDone();
+
+            if (purgedElementIds.Count == 0)
+                throw new InvalidOperationException(Properties.Resources.NoElementsToPurge);
+
+            return purgedElementIds.Select(x => x.IntegerValue).ToList();
+        }
+
+        private List<ElementId> PurgeMaterials()
+        {
+            List<Autodesk.Revit.DB.ElementId> materials = new FilteredElementCollector(this.InternalDocument)
+                .OfClass(typeof(Autodesk.Revit.DB.Material))
+                .ToElementIds()
+                .ToList();
+
+            if (materials == null)
+                return new List<ElementId>();
+
+            List<Autodesk.Revit.DB.Element> elementTypes = new FilteredElementCollector(this.InternalDocument)
+                .WhereElementIsElementType()
+                .ToElements()
+                .ToList();
+
+            if (elementTypes == null)
+                return new List<ElementId>();
+
+            var nonPurgedElements = new HashSet<ElementId>();
+
+            for (int i = 0; i < elementTypes.Count; i++)
+            {
+                nonPurgedElements.UnionWith(GetNonPurgableMaterialIds(elementTypes[i]));
+            }
+
+            List<ElementId> purgableMaterials = materials.Except(nonPurgedElements).ToList();
+            if (purgableMaterials.Count > 0)
+                this.InternalDocument.Delete(purgableMaterials);
+
+            List<ElementId> purgedMaterials = new List<ElementId>();
```

Comment:
think we can do same thing as method below it and inline this, eliminating allocation for a whole new list.